source/communication_interfaces/include/communication_interfaces/sockets/ISocket.hpp

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <state_representation/parameters/ParameterMap.hpp>
+
+#include "communication_interfaces/ByteArray.hpp"
+
+namespace communication_interfaces::sockets {
+
+/**
+ * @brief Interface class to define functions required for simple socket communication
+ */
+class ISocket : public state_representation::ParameterMap {
+public:
+  /**
+   * @brief Default constructor
+   */
+  ISocket() = default;
+
+  virtual ~ISocket() = default;
+
+  /**
+   * @brief Perform configuration steps for socket communication
+   */
+  virtual void configure() = 0;
+
+  /**
+   * @brief Receive bytes from the socket
+   * @param buffer The buffer to fill with the received bytes
+   * @return True if bytes were received, false otherwise
+   */
+  virtual bool receive_bytes(ByteArray& buffer) = 0;
+
+  /**
+   * @brief Send bytes to the socket
+   * @param buffer The buffer with the bytes to send
+   * @return True if bytes were sent, false otherwise
+   */
+  virtual bool send_bytes(const ByteArray& buffer) = 0;
+
+  /**
+   * @brief Perform steps to disconnect and close the socket communication
+   */
+  virtual void close() {}
```

Comment:
comment: it seems to me that `close()` should be generally called on destruction. However, destructors should not call virtual methods. You will need to be explicit about this behavior in derived classes. Either implement closing behavior in a non-virtual `on_close()` function that is called both in `close override` and the destructor, or define the closing logic in derived `close() final` and preferably be explicit with static invocation rather than dynamic `this->close()`. 

```c++
UDPClient::~UDPClient {
  UDPClient::close();
}

void UDPClient::close final {
  ...
}
```

For whichever approach you prefer, it may be prudent to leave a note to future developers (e.g. you and I) that we do it consistently and safely.

Reply:
I see the problem that your describing but frankly, I'm not sure to understand what to do. Would another solution just be to make the `close()` non-virtual? In both cases that you describe, we would still have virtual classes and hence we have to be careful not to call them from the destructor, but there is no real way to prohibit it no? I guess what I mean is that it seems it will always come down to the derived class to be implemented correctly so why not just say that `close()` has to be overriden?